File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1 or not l2: return l1 or l2
        nh = l1 if l1.val <= l2.val else l2
        dummy = ListNode(-1, nh)
        p1, p2 = l1, l2
        tail = dummy
        
        while p1 and p2:
            if p1.val <= p2.val:
                tail.next = p1
                p1 = p1.next
            else:
                tail.next = p2
                p2 = p2.next
            tail = tail.next
                
        if p1:
            tail.next = p1
        if p2:
            tail.next = p2
            
        return dummy.next
                
            
        

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode p = list1, q = list2;

        ListNode dummy = new ListNode(-1);

        ListNode tail = dummy;

        while (p != null || q != null) {
            int pval = p != null ? p.val : Integer.MAX_VALUE;
            int qval = q != null ? q.val : Integer.MAX_VALUE;

            if (pval <= qval) {
                ListNode nex = p.next;
                tail.next = p;
                tail = tail.next;
                tail.next = null;
                p = nex;
            } else {
                ListNode nex = q.next;
                tail.next = q;
                tail = tail.next;
                tail.next = null;
                q = nex;
            }
        }

        return dummy.next;

        
    }
}